---
title: 'SpringBoot transaction大坑'
date: '2025-06-10'
tags: ['Java', 'SpringBoot']
draft: false
---

Springboot trasaction 没有rollback,在抛出IOException之后


以下是我追踪到的路径(调用链条自下而上):

```java
	public boolean rollbackOn(Throwable ex) {
		return (ex instanceof RuntimeException || ex instanceof Error);
	}
```


```java

		if (winner == null) {
			return super.rollbackOn(ex);
		}
```

```java
	protected void completeTransactionAfterThrowing(@Nullable TransactionInfo txInfo, Throwable ex) {
		if (txInfo != null && txInfo.getTransactionStatus() != null) {
			if (logger.isTraceEnabled()) {
				logger.trace("Completing transaction for [" + txInfo.getJoinpointIdentification() +
						"] after exception: " + ex);
			}
			if (txInfo.transactionAttribute != null && txInfo.transactionAttribute.rollbackOn(ex)) {
                // ...
            }
            // ...
        }
    }
```

spring: 5.3.20, springboot: 2.7.0

https://docs.spring.io/spring-framework/docs/5.3.20/reference/html/data-access.html#transaction-declarative-rolling-back

没找到全局配置的方法，spring做得很坚决：

https://docs.spring.io/spring-framework/docs/5.3.20/javadoc-api/org/springframework/transaction/interceptor/DefaultTransactionAttribute.html#rollbackOn-java.lang.Throwable-

```
	protected TransactionAttribute parseTransactionAnnotation(AnnotationAttributes attributes) {
```
看这里，transactional的rollbackfor从这里进去，transactionAttribute.rollbackOn除了这里进去的还有两个写死的，就是runxxxx和error。

这样写就行了： 

```
@Transactional(rollbackFor = Exception.class)
```